include/hybridmap/HybridNeutralAtomMapper.hpp

```diff
@@ -0,0 +1,534 @@
+};
+
+/**
+ * @brief Helper struct to store the position of a multi qubit gate and the
+ * number of moves needed to execute it.
+ */
+struct MultiQubitMovePos {
+  CoordIndices coords;
+  size_t       nMoves{0};
+
+  // copy constructor
+  MultiQubitMovePos()                               = default;
+  MultiQubitMovePos(const MultiQubitMovePos& other) = default;
+
+  // copy assignment
+  MultiQubitMovePos& operator=(const MultiQubitMovePos& other) = default;
+};
+
+/**
+ * @brief Class to map a quantum circuit to a neutral atom architecture.
+ * @details The mapping has following important parts:
+ * - initial mapping: The initial mapping of the circuit qubits to the hardware
+ * qubits.
+ * - layer creation: The creation of the front and lookahead layers, done one
+ * the fly and taking into account basic commutation rules.
+ * - estimation: The estimation of the number of swap gates and moves needed to
+ * execute a given gate and the decision which technique is better.
+ * - gate based mapping: SABRE based algorithm to choose the bast swap for the
+ * given layers.
+ * - shuttling based mapping: Computing and evaluation of possible moves and
+ * choosing best.
+ * - multi-qubit-gates: Additional steps and checks to bring multiple qubits
+ * together.
+ * -> Final circuit contains abstract SWAP gates and MOVE operations, which need
+ * to be decomposed using AODScheduler.
+ */
+class NeutralAtomMapper {
+protected:
+  // The considered architecture
+  qc::NeutralAtomArchitecture arch;
+  // The mapped quantum circuit
+  qc::QuantumComputation mappedQc;
+  // The mapped quantum circuit converted to AOD movements
+  qc::QuantumComputation mappedQcAOD;
+  // The scheduler to schedule the mapped quantum circuit
+  NeutralAtomScheduler scheduler;
+  // The gates that have been executed
+  std::vector<const Operation*> executedCommutingGates;
+  // Gates in the front layer to be executed with swap gates
+  GateList frontLayerGate;
+  // Gates in the front layer to be executed with move operations
+  GateList frontLayerShuttling;
+  // Gates in the lookahead layer to be executed with swap gates
+  GateList lookaheadLayerGate;
+  // Gates in the lookahead layer to be executed with move operations
+  GateList lookaheadLayerShuttling;
+  // The minimal weight for any multi-qubit gate
+  fp twoQubitSwapWeight = 1;
+  // The runtime parameters of the mapper
+  MapperParameters parameters;
+  // The qubits that are blocked by the last swap
+  std::deque<std::set<HwQubit>> lastBlockedQubits;
+  // The last moves that have been executed
+  std::deque<AtomMove> lastMoves;
+  // Precomputed decay weights
+  std::vector<fp> decayWeights;
+  // Counter variables
+  uint32_t nSwaps = 0;
+  uint32_t nMoves = 0;
+
+  // The current placement of the hardware qubits onto the coordinates
+  HardwareQubits hardwareQubits;
+  // The current mapping between circuit qubits and hardware qubits
+  Mapping mapping;
+
+  bool verbose = false;
```

Comment:
Shouldn't that also be a MapperParameter?

Reply:
Added it to the parameters